src/library/PiaFormat.ts

```diff
@@ -0,0 +1,206 @@
+import * as dayjs from "dayjs";
+import data from "src/pages/data";
+
+enum PIAGender {
+  male = 0,
+  female = 1,
+}
+
+class PIAData {
+  ssn?: string;
+  birthDate?: Date;
+  sex?: PIAGender;
+
+  assign(key: keyof PIAData, value: any) {
+    switch (key) {
+      case "ssn":
+        this.ssn = value;
+        break;
+      case "birthDate":
+        this.birthDate = value;
+        break;
+      case "sex":
+        this.sex = value;
+        break;
+    }
+  }
+}
+
+enum PIATypes {
+  piaString = "string",
+  piaGender = "piaGender",
+  piaDate = "date",
+  piaMonthYear = "monthyear",
+  piaInt = "int",
+  piaDecimal = "decimal",
+  piaFloat = "float",
+}
+
+class PIAFieldFormat {
+  name: keyof PIAData;
+  start: number;
+  end: number;
+  type: PIATypes;
+  constructor(name: keyof PIAData, start: number, end: number, type: PIATypes) {
+    this.name = name;
+    this.start = start;
+    this.end = end;
+    this.type = type;
+  }
+}
+const piaLineFormats: Record<string, PIAFieldFormat[]> = {
+  "01": [
+    new PIAFieldFormat("ssn", 3, 11, PIATypes.piaString),
```

Comment:
Very declarative! But for Typescript to help us here, we need a mapping from PIA type to the target parsed type (e.g., `PIATypes.String -> string`). Basically, we need to make sure that Typescript can check that `birthDate` gets parsed to a date and not a string.

Luckily, we can just type the constructor a little more strictly:

```
type PIAParser<T extends keyof PIAData> = (input: string) => PIAData[T]
class PIAFieldFormat<T extends keyof PIAData> {
  constructor(name: T, start: number, end: number, parse: PIAParser<T>) {
    // ...
  }
}
```

The following will fail:
```
new PIAFieldFormat("ssn", 3, 11, parsePiaDate) // Date cannot be assigned to string
```

This way, we can just call `fieldFormat.parse(input)`, and we're guaranteed to get a value that corresponds to a key in `PIAData` _and_ has the correct type. We can probably drop `PIATypes` entirely too!